src/Expander.hs

```diff
@@ -1473,8 +1465,24 @@ expandOneForm prob stx
           Id _ -> error "Impossible happened - identifiers are identifier-headed!"
 
 
-expandOneDeclaration :: Scope -> DeclPtr -> Syntax -> DeclValidityPtr -> Expand ()
-expandOneDeclaration sc dest stx ph = expandOneForm (DeclDest dest sc ph) stx
+expandDeclForm :: DeclTreePtr -> ScopeSet -> DeclValidityPtr -> Syntax -> Expand ()
+expandDeclForm dest scs vp stx =
+  expandOneForm (DeclTreeDest dest scs vp) stx
+
+expandDeclForms :: DeclTreePtr -> ScopeSet -> DeclValidityPtr -> Syntax -> Expand ()
+expandDeclForms dest scs vp (Syntax (Stx _ _ (List []))) = do
+  linkDeclTree dest DeclTreeLeaf
+  linkDeclValidity vp scs
+expandDeclForms dest scs vp (Syntax (Stx stxScs loc (List (d:ds)))) = do
+  carDest <- liftIO $ newDeclTreePtr
+  cdrDest <- liftIO $ newDeclTreePtr
+  carValidityPtr <- newDeclValidityPtr
+  linkDeclTree dest (DeclTreeBranch carDest cdrDest)
+  forkExpandDeclForms cdrDest carValidityPtr vp
+    (Syntax (Stx stxScs loc (List ds)))
+  expandDeclForm carDest scs carValidityPtr =<< addRootScope d
```

Comment:
This is where the `ScopeSet` is threaded through: from the input `scs`, to the `car`'s `expandDeclForm` task, who adds some `Scope`s to it and writes the result to `carValidityPtr`, which is read (and blocked on) by the `forkExpandDeclForms` task, who adds even more `Scope`s and writes the result to `vp`.

Reply:
What about adding this text as a comment in the source file?

Reply:
good idea, added!